qir/qat/GroupingPass/GroupingPass.cpp

```diff
@@ -19,27 +19,106 @@ namespace quantum
         return true;
     }
 
-    void GroupingPass::prepareBlockModification(llvm::Module& module, llvm::BasicBlock* tail_classical)
+    bool GroupingPass::isQuantumRegister(llvm::Type const* type)
+    {
+        if (type->isPointerTy())
+        {
+            auto element_type = type->getPointerElementType();
+            if (element_type->isStructTy())
+            {
+                auto type_name = static_cast<String>(element_type->getStructName());
+                return quantum_register_types_.find(type_name) != quantum_register_types_.end();
+            }
+        }
```

Comment:
I didn't dive deep in the code but just in case I would like to make sure that you have already considered placing `const` near every `auto`. E.g.
```suggestion
        if (type->isPointerTy())
        {
            auto const element_type = type->getPointerElementType();
            if (element_type->isStructTy())
            {
                auto const type_name = static_cast<String const>(element_type->getStructName());
                return quantum_register_types_.find(type_name) != quantum_register_types_.end();
            }
        }
```